libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,363 @@
+  startUpdate: FirstUpdate,
+) {
+  const { data, encounteredIds } = startUpdate();
+
+  switch (environment.store.kind) {
+    case 'BaseNode': {
+      mergeDataLayer(environment.store.data, data);
+      break;
+    }
+    case 'StartUpdateNode': {
+      const prevStartUpdate = environment.store.startUpdate;
+
+      mergeDataLayer(environment.store.data, data);
+
+      environment.store.startUpdate = () => {
+        const { data } = prevStartUpdate();
+        mergeDataLayer(data, startUpdate().data);
+        return { data };
+      };
+
+      break;
+    }
+    case 'NetworkResponseNode':
+    case 'OptimisticNode': {
+      const node: StartUpdateNode = {
+        kind: 'StartUpdateNode',
+        parentNode: environment.store,
+        childNode: null,
+        data,
+        startUpdate: startUpdate,
+      };
+      environment.store.childNode = node;
+      environment.store = node;
+      break;
+    }
+    default: {
+      environment.store satisfies never;
+    }
+  }
+
+  logMessage(environment, () => ({
+    kind: 'StartUpdateComplete',
+    updatedIds: encounteredIds,
+  }));
+
+  callSubscriptions(environment, encounteredIds);
+}
+
+export function addOptimisticNode(
+  environment: IsographEnvironment,
+  startUpdate: FirstUpdate,
+) {
+  const { data, encounteredIds } = startUpdate();
+
+  switch (environment.store.kind) {
+    case 'BaseNode':
+    case 'StartUpdateNode':
+    case 'NetworkResponseNode':
+    case 'OptimisticNode': {
+      const node: OptimisticNode = {
+        kind: 'OptimisticNode',
+        parentNode: environment.store,
+        childNode: null,
+        data,
+        startUpdate: startUpdate,
+      };
+
+      environment.store.childNode = node;
+      environment.store = node;
+
+      callSubscriptions(environment, encounteredIds);
+      return (data: DataLayer) => {
+        const encounteredIds: EncounteredIds = new Map();
+        compareData(node.data, data, encounteredIds);
+        replaceOptimisticNodeWithNetworkResponseNode(
+          environment,
+          node,
+          data,
+          encounteredIds,
+        );
+        callSubscriptions(environment, encounteredIds);
+      };
+    }
+    default: {
+      environment.store satisfies never;
+      throw new Error('Unreachable');
+    }
+  }
+}
+
+function mergeParentNodes(
+  environment: IsographEnvironment,
+  node: OptimisticNode | NetworkResponseNode | StartUpdateNode | null,
+  mutableEncounteredIds: EncounteredIds,
+) {
+  while (node && node?.kind !== 'OptimisticNode') {
+    const data = 'startUpdate' in node ? node.startUpdate().data : node.data;
+    compareData(node.data, data, mutableEncounteredIds);
+    mergeDataLayer(environment.store.data, data);
+    node = node.childNode;
+  }
+  return node;
+}
+
+function reexecuteUpdates(
+  environment: IsographEnvironment,
+  node: OptimisticNode | NetworkResponseNode | StartUpdateNode | null,
+  mutableEncounteredIds: EncounteredIds,
+) {
+  while (node !== null) {
```

Comment:
Do we need to reassign `environment.store` here? Can we just go from `env.store -> env.store.parent` etc until we encounter null?